A33-Fault-Injection.md

```diff
@@ -0,0 +1,253 @@
+
+```bash
+grpc-fault-injection-delay-ms=1.5       # 1.5ms delay before sending first request
+grpc-fault-injection-delay-percent=0.2  # This call has 20% chance to be delayed
+grpc-fault-injection-abort-code=7       # Reject the call as PERMISSION_DENIED
+grpc-fault-injection-abort-percent=0.5  # This call has 50% chance to be rejected
+```
+
+In this way, users can utilize fault injection without configuring xDS. It also allows decouple the enforcement of FI from xDS control path.
+
+But similar to above approach, adding extra metadata is not a good idea.
+
+
+### Alternative 3: Fault Injection Config via ServiceConfig
+
+Across Core, Java and Go, Service Config is a mechanism to supply advanced configuration for specific service or method. It’s natural to add a new section of fault injection policy into the MethodConfig. The proposed proto configuration looks like:
+
+```protobuf
+message MethodConfig {
+
+ // ...ignoring other fields
+
+ // The fault injection policy for outbound RPCs.
+ //
+ // The fault injection offers delay injection and aborting RPC functionality.
+ // Both of them can be specified at the same time. If both features are
+ // triggered, the RPC will be delayed first then abort with the status code.
+ message FaultInjectionPolicy {
+   // The fixed delay before starting the RPC.
+   //
+   // No traffic should be sent until the delay phase ends. By default, the
+   // delay duration is 0 (0 seconds and 0 nanos). If the delay is 0, the RPC
+   // delay feature is disabled and delay_fraction will be ignored.
+   google.protobuf.Duration delay = 1;
+
+   // The probability that this RPC will be delayed.
+   //
+   // The default value is 0. The probability is represented as a fraction over
+   // 1 million (1000000). For example, 0.625% will be 6250, 12.5% will be
+   // 125000. Its value should be clamped into the interval [0, 1000000].
+   uint32 delay_fraction = 2;
+
+   // The RPC status (code and message) for the aborted RPC.
+   //
+   // By default the status code is 0 (OK), which disables the RPC abort
+   // feature and ignores abort_percent. If the control plane only offers
+   // HTTP status code, the application should map it to cardinal RPC status
+   // code.
+   //
+   // By default the status message is empty. If the status message is
+   // specified with non-OK status code, then the status message will be added
+   // to the aborted RPC.
+   google.rpc.Status abort_status = 3;
+
+   // The probability that this RPC will be aborted.
+   //
+   // The default value is 0. The probability is represented as a fraction over
+   // 1 million (1000000). For example, 0.625% will be 6250, 12.5% will be
+   // 125000. Its value should be clamped into the interval [0, 1000000].
+   uint32 abort_fraction = 4;
+ }
+
+ FaultInjectionPolicy fault_injection_policy = 8;
+}
+```
+
+This approach requires more engineering effort, but provides the fault injection feature to non-xDS gRPC applications.
```

Comment:
Might be worth noting that we may consider doing something like this in the future.